src/NuGet.Clients/NuGet.SolutionRestoreManager/SolutionRestoreCommand.cs

```diff
@@ -103,25 +105,23 @@ internal sealed class SolutionRestoreCommand
         /// <param name="e">Event args.</param>
         private void OnRestorePackages(object sender, EventArgs args)
         {
-            if (!SolutionRestoreWorker.IsBusy)
-            {
-                SolutionRestoreWorker.Restore(SolutionRestoreRequest.ByMenu());
-            }
-            else
+            if (_restoreTask.IsCompleted)
             {
-                // QueryStatus should disable the context menu in most of the cases.
-                // Except when NuGetPackage was not loaded before VS won't send QueryStatus.
-                Logger.Log(MessageLevel.Info, Resources.SolutionRestoreFailed_RestoreWorkerIsBusy);
+                _restoreTask = NuGetUIThreadHelper.JoinableTaskFactory
+                    .RunAsync(() => SolutionRestoreWorker.ScheduleRestoreAsync(
+                        SolutionRestoreRequest.ByMenu(),
+                        CancellationToken.None))
+                    .Task;
```

Comment:
The worker itself will take care of cancelling the task. Tested this particular scenario manually by closing the solution while restore is in progress.